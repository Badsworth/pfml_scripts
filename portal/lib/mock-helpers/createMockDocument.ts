import {
  BenefitsApplicationDocument,
  DocumentType,
} from "../../src/models/Document";

export const createMockBenefitsApplicationDocument = (
  customAttrs: Partial<BenefitsApplicationDocument> = {}
): BenefitsApplicationDocument => ({
  application_id: "mock_application_id",
  content_type: "application/pdf",
  created_at: "2020-01-01T00:00:00.000Z",
  description: "",
  document_type:
    DocumentType.certification["Serious Health Condition - Employee"],
  fineos_document_id: "mock_document_id",
  name: "mock_document_name",
  user_id: "mock_user_id",
  ...customAttrs,
});
