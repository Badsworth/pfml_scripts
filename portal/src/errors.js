/**
 * @file Custom Error classes. Useful as a way to see all potential errors that our system may throw/catch
 */

class BasePortalError extends Error {
  constructor(...params) {
    super(...params);

    // Maintains proper stack trace for where our error was thrown in modern browsers
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, BasePortalError);
    }
  }
}

/**
 * A fetch request failed due to a network error. The error wasn't the fault of the user,
 * and an issue was encountered while setting up or sending a request, or parsing the response.
 * Examples of a NetworkError could be the user's device lost internet connection, a CORS issue,
 * or a malformed request.
 */
export class NetworkError extends BasePortalError {
  constructor(...params) {
    super(...params);
    this.name = "NetworkError";
  }
}

/**
 * A fetch request failed due to a 404 error
 */
export class NotFoundError extends BasePortalError {
  constructor(...params) {
    super(...params);
    this.name = "NotFoundError";
  }
}

/**
 * A transition between the current route state to the next route failed most likely because
 * a CONTINUE transition was not defined for the current route.
 */
export class RouteTransitionError extends BasePortalError {
  constructor(...params) {
    super(...params);
    this.name = "RouteTransitionError";
  }
}

/**
 * users/current api resource did not return a user in its response
 */
export class UserNotReceivedError extends BasePortalError {
  constructor(...params) {
    super(...params);
    this.name = "UserNotReceivedError";
  }
}

/**
 * There was an error when attempting to a request to the users/current api resource
 */
export class UserNotFoundError extends BasePortalError {
  constructor(...params) {
    super(...params);
    this.name = "UserNotFoundError";
  }
}

/**
 * An API response returned a status code greater than 400
 */
export class ApiRequestError extends BasePortalError {
  constructor(...params) {
    super(...params);
    this.name = "ApiRequestError";
  }
}

/**
 * An API response returned a 400 status code and its JSON body didn't include any `errors`
 */
export class BadRequestError extends ApiRequestError {
  constructor(...params) {
    super(...params);
    this.name = "BadRequestError";
  }
}

/**
 * An API response returned a 403 status code, indicating an issue with the authorization of the request.
 * Examples of a ForbiddenError could be the user's browser prevented the session cookie from
 * being created, or the user hasn't consented to the data sharing agreement.
 */
export class ForbiddenError extends ApiRequestError {
  constructor(...params) {
    super(...params);
    this.name = "ForbiddenError";
  }
}

/**
 * An API response returned a 500 status code
 */
export class InternalServerError extends ApiRequestError {
  constructor(...params) {
    super(...params);
    this.name = "InternalServerError";
  }
}

/**
 * An API response returned a 408 status code
 */
export class RequestTimeoutError extends ApiRequestError {
  constructor(...params) {
    super(...params);
    this.name = "RequestTimeoutError";
  }
}

/**
 * A request to an Application's `/documents` endpoint failed
 */
export class DocumentsRequestError extends BasePortalError {
  constructor(application_id, ...params) {
    super(...params);
    this.application_id = application_id;
    this.name = "DocumentsRequestError";
  }
}

/**
 *  An API response returned a 503 status code
 */
export class ServiceUnavailableError extends ApiRequestError {
  constructor(...params) {
    super(...params);
    this.name = "ServiceUnavailableError";
  }
}

/**
 * An API response returned a 401 status code
 */
export class UnauthorizedError extends ApiRequestError {
  constructor(...params) {
    super(...params);
    this.name = "UnauthorizedError";
  }
}

/**
 * A request wasn't completed due to one or more validation issues
 */
export class ValidationError extends BasePortalError {
  /**
   * @param {{ field: string, message: string, rule: string, type: string }[]} issues - List of validation issues returned by the API
   * @param {string} i18nPrefix - Used in the i18n message keys, prefixed to the field name (e.g. `prefix.field_name`)
   * @example new ValidationError([{ field: "tax_identifier", type: "pattern", message: "Field didn't match \d{9}" }], "claims")
   */
  constructor(issues, i18nPrefix, ...params) {
    super(...params);
    this.issues = issues;
    this.i18nPrefix = i18nPrefix;
    this.name = "ValidationError";
  }
}
