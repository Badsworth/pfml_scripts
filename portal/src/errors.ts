/**
 * @file Custom Error classes. Useful as a way to see all potential errors that our system may throw/catch
 */

class BasePortalError extends Error {
  constructor(...params) {
    super(...params);

    // Maintains proper stack trace for where our error was thrown in modern browsers
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, BasePortalError);
    }
  }
}

/**
 * A Cognito authentication step failed.
 */
export class CognitoAuthError extends BasePortalError {
  /**
   * @param {{ code: string, name: string, message: string }} cognitoError - error returned by cognito API
   * @param {{ type: string, field: string, message: string }} issue
   */
  constructor(cognitoError, issue = null, ...params) {
    super(...params);
    this.name = "CognitoAuthError";
    this.cognitoError = cognitoError;
    this.issue = issue;
  }
}

/**
 * The authenticated user's session expired or couldn't be found.
 */
export class AuthSessionMissingError extends BasePortalError {
  constructor(...params) {
    super(...params);
    this.name = "AuthSessionMissingError";
  }
}

/**
 * A fetch request failed due to a network error. The error wasn't the fault of the user,
 * and an issue was encountered while setting up or sending a request, or parsing the response.
 * Examples of a NetworkError could be the user's device lost internet connection, a CORS issue,
 * or a malformed request.
 */
export class NetworkError extends BasePortalError {
  constructor(...params) {
    super(...params);
    this.name = "NetworkError";
  }
}

/**
 * A transition between the current route state to the next route failed most likely because
 * a CONTINUE transition was not defined for the current route.
 */
export class RouteTransitionError extends BasePortalError {
  constructor(...params) {
    super(...params);
    this.name = "RouteTransitionError";
  }
}

/**
 * users/current api resource did not return a user in its response
 */
export class UserNotReceivedError extends BasePortalError {
  constructor(...params) {
    super(...params);
    this.name = "UserNotReceivedError";
  }
}

/**
 * An API response returned a status code greater than 400
 */
export class ApiRequestError extends BasePortalError {
  constructor(data, ...params) {
    super(...params);
    this.responseData = data;
    this.name = "ApiRequestError";
  }
}

/**
 * A fetch request failed due to a 404 error
 */
export class NotFoundError extends ApiRequestError {
  constructor(...params) {
    super(...params);
    this.name = "NotFoundError";
  }
}

/**
 * An API response returned a 400 status code and its JSON body didn't include any `errors`
 */
export class BadRequestError extends ApiRequestError {
  constructor(...params) {
    super(...params);
    this.name = "BadRequestError";
  }
}

/**
 * An API response returned a 403 status code, indicating an issue with the authorization of the request.
 * Examples of a ForbiddenError could be the user's browser prevented the session cookie from
 * being created, or the user hasn't consented to the data sharing agreement.
 */
export class ForbiddenError extends ApiRequestError {
  constructor(...params) {
    super(...params);
    this.name = "ForbiddenError";
  }
}

/**
 * An API response returned a 500 status code
 */
export class InternalServerError extends ApiRequestError {
  constructor(...params) {
    super(...params);
    this.name = "InternalServerError";
  }
}

/**
 * An API response returned a 408 status code
 */
export class RequestTimeoutError extends ApiRequestError {
  constructor(...params) {
    super(...params);
    this.name = "RequestTimeoutError";
  }
}

/**
 * A GET request to an Application's `/documents` endpoint failed
 */
export class DocumentsLoadError extends BasePortalError {
  /**
   * @param {string} application_id - ID of the Claim the documents are associated with
   * @example new DocumentsLoadError('mock_application_id')
   */
  constructor(application_id, ...params) {
    super(...params);
    this.application_id = application_id;
    this.name = "DocumentsLoadError";
  }
}

/**
 * A POST request to an Application's `/documents` endpoint failed
 */
export class DocumentsUploadError extends BasePortalError {
  /**
   * Since we construct DocumentsUploadError when we catch them in document logic,
   * the error could be a ValidationError or some other type of errors.
   * If it's a ValidationError, we need to pass the validation message.
   *
   * @param {string} application_id - ID of the Claim the documents are associated with
   * @param {string} file_id - ID of the file causing errors, so the issues can be displayed inline
   * @param {{ field: string, message: string, rule: string, type: string }} issue - the validation issue returned by the API
   * @example new DocumentsUploadError('mock_application_id',[{ field: "", type: "fineos", message: "File size limit" }])
   */
  constructor(application_id, file_id, issue = null, ...params) {
    super(...params);
    this.application_id = application_id;
    this.file_id = file_id;
    this.issue = issue;
    this.name = "DocumentsUploadError";
  }
}

/**
 *  An API response returned a 503 status code
 */
export class ServiceUnavailableError extends ApiRequestError {
  constructor(...params) {
    super(...params);
    this.name = "ServiceUnavailableError";
  }
}

/**
 * An API response returned a 401 status code
 */
export class UnauthorizedError extends ApiRequestError {
  constructor(...params) {
    super(...params);
    this.name = "UnauthorizedError";
  }
}

/**
 * A request wasn't completed due to one or more validation issues
 */
export class ValidationError extends BasePortalError {
  /**
   * @param {{ field: string, message: string, rule: string, type: string }[]} issues - List of validation issues returned by the API
   * @param {string} i18nPrefix - Used in the i18n message keys, prefixed to the field name (e.g. `prefix.field_name`)
   * @example new ValidationError([{ field: "tax_identifier", type: "pattern", message: "Field didn't match \d{9}" }], "applications")
   */
  constructor(issues, i18nPrefix, ...params) {
    super(...params);
    this.issues = issues;
    this.i18nPrefix = i18nPrefix;
    this.name = "ValidationError";
  }
}

/**
 * An error was encountered because the current user is not verified.
 */
export class LeaveAdminForbiddenError extends ForbiddenError {
  constructor(employer_id, has_verification_data, message, ...params) {
    super(...params);
    this.employer_id = employer_id;
    this.has_verification_data = has_verification_data;
    this.message = message;
    this.name = "LeaveAdminForbiddenError";
  }
}
