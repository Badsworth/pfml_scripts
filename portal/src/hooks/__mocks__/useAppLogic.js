import Claim from "../../models/Claim";
import ClaimCollection from "../../models/ClaimCollection";
import { uniqueId } from "lodash";

export default jest.fn(() => ({
  auth: {
    createAccount: jest.fn(),
    forgotPassword: jest.fn(),
    login: jest.fn(),
  },
  claims: new ClaimCollection(),
  createClaim: jest.fn(() => new Claim({ application_id: uniqueId() })),
  loadClaims: jest.fn(),
  submitClaim: jest.fn(),
  updateClaim: jest.fn(),
}));
