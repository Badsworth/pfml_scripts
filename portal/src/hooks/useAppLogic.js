/* eslint sort-keys: ["error", "asc"] */
import useAppErrorsLogic from "./useAppErrorsLogic";
import useAuthLogic from "./useAuthLogic";
import useBenefitsApplicationsLogic from "./useBenefitsApplicationsLogic";
import useClaimsLogic from "./useClaimsLogic";
import useDocumentsLogic from "./useDocumentsLogic";
import useEmployersLogic from "./useEmployersLogic";
import useFlagsLogic from "./useFeatureFlagsLogic";
import useOtherLeavesLogic from "./useOtherLeavesLogic";
import usePortalFlow from "./usePortalFlow";
import useUsersLogic from "./useUsersLogic";

const useAppLogic = () => {
  // Utility to determine application flow
  const portalFlow = usePortalFlow();

  // State representing currently visible errors and warnings
  const { appErrors, ...appErrorsLogic } = useAppErrorsLogic({ portalFlow });
  const auth = useAuthLogic({ appErrorsLogic, portalFlow });

  // State representing the Portal's user object.
  // Initialize to empty user but will be populated upon the first API call
  // to fetch the user (or create the user on their first login)
  const users = useUsersLogic({
    appErrorsLogic,
    isLoggedIn: auth.isLoggedIn,
    portalFlow,
  });

  // user will be eventually set through a `login` method
  // const [user, setUser] = useState();
  // TODO (CP-885): remove user from configs and write login method
  const benefitsApplications = useBenefitsApplicationsLogic({
    appErrorsLogic,
    portalFlow,
    user: users.user,
  });

  const claims = useClaimsLogic({
    appErrorsLogic,
  });

  const documents = useDocumentsLogic({
    appErrorsLogic,
    portalFlow,
  });

  const employers = useEmployersLogic({
    appErrorsLogic,
    portalFlow,
    setUser: users.setUser,
  });

  const otherLeaves = useOtherLeavesLogic({ appErrorsLogic });
  const featureFlags = useFlagsLogic({ appErrorsLogic });

  return {
    appErrors,
    auth,
    benefitsApplications,
    catchError: appErrorsLogic.catchError,
    claims,
    // TODO (CP-886): remove once all API calls are behind appLogic
    clearErrors: appErrorsLogic.clearErrors,
    documents,
    employers,
    otherLeaves,
    portalFlow,
    setAppErrors: appErrorsLogic.setAppErrors,
    users,
    featureFlags,
  };
};

export default useAppLogic;
