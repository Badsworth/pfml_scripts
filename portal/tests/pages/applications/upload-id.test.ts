import {
  BenefitsApplicationDocument,
  DocumentType,
} from "../../../src/models/Document";
import {
  MockBenefitsApplicationBuilder,
  makeFile,
  renderPage,
} from "../../test-utils";
import { act, screen, waitFor } from "@testing-library/react";
import ApiResourceCollection from "src/models/ApiResourceCollection";
import AppErrorInfo from "../../../src/models/AppErrorInfo";
import AppErrorInfoCollection from "../../../src/models/AppErrorInfoCollection";
import { AppLogic } from "../../../src/hooks/useAppLogic";
import UploadId from "../../../src/pages/applications/upload-id";
import { ValidationError } from "../../../src/errors";
import { createMockBenefitsApplicationDocument } from "../../../lib/mock-helpers/createMockDocument";
import { setupBenefitsApplications } from "../../test-utils/helpers";
import userEvent from "@testing-library/user-event";

jest.mock("../../../src/hooks/useAppLogic");
jest.mock("../../../src/services/tracker");

const expectedTempFiles = [
  expect.objectContaining({
    file: expect.objectContaining({ name: "file1" }),
  }),
  expect.objectContaining({
    file: expect.objectContaining({ name: "file2" }),
  }),
];

const cbHasLoadedClaimDocsTrue = (appLogic: AppLogic) => {
  appLogic.documents.hasLoadedClaimDocuments = jest.fn().mockReturnValue(true);
};

const clearErrors = jest.fn();
const catchError = jest.fn();

const goToNextPage = jest.fn(() => {
  return Promise.resolve();
});

let attach = jest.fn();

const setup = (
  claim = new MockBenefitsApplicationBuilder().medicalLeaveReason().create(),
  props = {},
  cb?: (appLogic: AppLogic) => void
) => {
  return renderPage(
    UploadId,
    {
      addCustomSetup: (appLogic) => {
        setupBenefitsApplications(appLogic, [claim], cb);
        appLogic.portalFlow.goToNextPage = goToNextPage;
        appLogic.documents.attach = attach;
        appLogic.clearErrors = clearErrors;
        appLogic.catchError = catchError;
      },
    },
    { query: { claim_id: "mock_application_id" }, ...props }
  );
};

describe("UploadId", () => {
  describe("before any documents have been loaded", () => {
    beforeEach(() => {
      setup();
    });

    it("does not render a FileCardList", () => {
      expect(screen.queryByText(/File/)).not.toBeInTheDocument();
    });

    it("renders a spinner", () => {
      expect(screen.getByRole("progressbar")).toBeInTheDocument();
    });
  });

  describe("after documents have been loaded", () => {
    it("when the user doesn't have a Mass ID, renders the page with 'Other' ID content", () => {
      const claim = new MockBenefitsApplicationBuilder()
        .medicalLeaveReason()
        .create();
      claim.has_state_id = false;
      const { container } = setup(claim, undefined, cbHasLoadedClaimDocsTrue);
      expect(container).toMatchSnapshot();
    });

    describe("when the user has a Mass ID", () => {
      const cb = (appLogic: AppLogic) => {
        appLogic.documents.hasLoadedClaimDocuments = jest
          .fn()
          .mockReturnValue(true);
        appLogic.documents.documents =
          new ApiResourceCollection<BenefitsApplicationDocument>(
            "fineos_document_id",
            [
              createMockBenefitsApplicationDocument({
                document_type: DocumentType.identityVerification,
              }),
            ]
          );
      };
      beforeEach(() => {
        const claim = new MockBenefitsApplicationBuilder()
          .medicalLeaveReason()
          .create();
        claim.has_state_id = true;
        setup(claim, undefined, cb);
      });

      it("renders a FileCardList", () => {
        expect(screen.queryByText(/File/)).toBeInTheDocument();
      });

      it("doesn't render a spinner", () => {
        expect(screen.queryByRole("progressbar")).not.toBeInTheDocument();
      });

      it("renders the page with Mass ID content", () => {
        expect(
          screen.getByText(
            /Upload the front and back of your Massachusetts driver’s license or ID card/
          )
        ).toBeInTheDocument();
        expect(
          screen.getByRole("heading", { name: "File 1" })
        ).toBeInTheDocument();
        expect(screen.getByText("Choose files")).toBeInTheDocument();
      });
    });

    it("when the user uploads files, renders filecards for the files", async () => {
      setup(undefined, undefined, cbHasLoadedClaimDocsTrue);
      const tempFiles = [makeFile(), makeFile(), makeFile()];
      await act(async () => {
        await userEvent.upload(
          screen.getByLabelText(/Choose files/),
          tempFiles
        );
      });
      await waitFor(() => {
        expect(screen.getAllByText(/File/)).toHaveLength(3);
      });
    });

    it("clears errors", async () => {
      setup(undefined, undefined, cbHasLoadedClaimDocsTrue);
      const tempFiles = [makeFile(), makeFile(), makeFile()];
      await act(async () => {
        await userEvent.upload(
          screen.getByLabelText(/Choose files/),
          tempFiles
        );
      });

      expect(clearErrors).toHaveBeenCalledTimes(1);
    });

    it("catches invalid files", async () => {
      setup(undefined, undefined, cbHasLoadedClaimDocsTrue);
      const invalidFiles = [makeFile({ name: "file1", type: "image/heic" })];
      await act(async () => {
        await userEvent.upload(
          screen.getByLabelText(/Choose files/),
          invalidFiles
        );
      });

      await waitFor(() => {
        expect(catchError).toHaveBeenCalledWith(
          new ValidationError(
            [
              {
                field: "file",
                message:
                  "We could not upload: file1. Choose a PDF or an image file (.jpg, .jpeg, .png).",
                type: "required",
              },
            ],
            "documents"
          )
        );
      });
    });

    it("makes documents.attach request when no documents exist", async () => {
      const claim = new MockBenefitsApplicationBuilder().create();
      setup(claim, undefined, cbHasLoadedClaimDocsTrue);

      const tempFiles = [
        makeFile({ name: "file1" }),
        makeFile({ name: "file2" }),
      ];

      await act(async () => {
        await userEvent.upload(
          screen.getByLabelText(/Choose files/),
          tempFiles
        );
      });

      userEvent.click(
        screen.getByRole("button", { name: "Save and continue" })
      );

      await waitFor(() => {
        expect(attach).toHaveBeenCalledWith(
          claim.application_id,
          expect.arrayContaining(expectedTempFiles),
          expect.any(String),
          false
        );
      });
    });

    it("navigates the user to the next page if there are no errors", async () => {
      const tempFiles = [makeFile(), makeFile(), makeFile()];
      attach = jest.fn().mockImplementation(
        jest.fn(() => {
          return [
            Promise.resolve({ success: true }),
            Promise.resolve({ success: true }),
            Promise.resolve({ success: true }),
          ];
        })
      );
      setup(undefined, undefined, cbHasLoadedClaimDocsTrue);

      await act(async () => {
        await userEvent.upload(
          screen.getByLabelText(/Choose files/),
          tempFiles
        );
      });

      expect(
        screen.getAllByRole("button", { name: "Remove file" })[0]
      ).toBeEnabled();
      userEvent.click(
        screen.getByRole("button", { name: "Save and continue" })
      );
      expect(
        screen.getAllByRole("button", { name: "Remove file" })[0]
      ).toBeDisabled();

      await waitFor(() => {
        expect(goToNextPage).toHaveBeenCalled();
      });
    });

    it("does not navigate the user to the next page if there are errors", async () => {
      attach = jest.fn().mockImplementation(
        jest.fn(() => {
          return [Promise.resolve({ success: false })];
        })
      );
      const tempFiles = [makeFile()];
      setup(undefined, undefined, cbHasLoadedClaimDocsTrue);

      await act(async () => {
        await userEvent.upload(
          screen.getByLabelText(/Choose files/),
          tempFiles
        );
      });

      userEvent.click(
        screen.getByRole("button", { name: "Save and continue" })
      );
      await waitFor(() => {
        expect(goToNextPage).not.toHaveBeenCalled();
      });
    });

    it("displays successfully uploaded files as unremovable file cards", async () => {
      const claim = new MockBenefitsApplicationBuilder()
        .medicalLeaveReason()
        .create();

      attach = jest.fn().mockImplementation(
        jest.fn(() => {
          return [
            Promise.resolve({ success: true }),
            Promise.resolve({ success: true }),
            Promise.resolve({ success: true }),
          ];
        })
      );

      const tempFiles = [
        makeFile({ name: "File1" }),
        makeFile({ name: "File2" }),
        makeFile({ name: "File3" }),
      ];

      setup(claim, undefined, cbHasLoadedClaimDocsTrue);

      await act(async () => {
        await userEvent.upload(
          screen.getByLabelText(/Choose files/),
          tempFiles
        );
      });
      await waitFor(() => {
        expect(screen.getAllByText(/File /)).toHaveLength(3);
        expect(
          screen.queryByText(/You can’t remove files previously uploaded./)
        ).not.toBeInTheDocument();
      });

      userEvent.click(
        screen.getByRole("button", { name: "Save and continue" })
      );

      await waitFor(() => {
        expect(screen.queryByText(/File [0-9]/)).not.toBeInTheDocument();
      });
    });

    it("displays unsucessfully uploaded files as removable file cards", async () => {
      const claim = new MockBenefitsApplicationBuilder()
        .medicalLeaveReason()
        .create();
      attach = jest.fn().mockImplementation(
        jest.fn(() => {
          return [
            Promise.resolve({ success: true }),
            Promise.resolve({ success: true }),
            Promise.resolve({ success: false }),
          ];
        })
      );

      setup(claim, undefined, cbHasLoadedClaimDocsTrue);

      const tempFiles = [
        makeFile({ name: "File1" }),
        makeFile({ name: "File2" }),
        makeFile({ name: "File3" }),
      ];

      await act(async () => {
        await userEvent.upload(
          screen.getByLabelText(/Choose files/),
          tempFiles
        );
      });

      await waitFor(() => {
        expect(
          screen.getAllByRole("heading", { name: /File [0-9]/ })
        ).toHaveLength(3);
        expect(
          screen.queryByText(/You can’t remove files previously uploaded./)
        ).not.toBeInTheDocument();
      });

      userEvent.click(
        screen.getByRole("button", { name: "Save and continue" })
      );
      await waitFor(() => {
        expect(
          screen.getByRole("heading", { name: /File/ })
        ).toBeInTheDocument();
      });
    });

    it("redirects to the Applications page when the claim has been completed", async () => {
      const claim = new MockBenefitsApplicationBuilder()
        .medicalLeaveReason()
        .create();
      attach = jest.fn().mockImplementation(
        jest.fn(() => {
          return [
            Promise.resolve({ success: true }),
            Promise.resolve({ success: true }),
          ];
        })
      );

      setup(claim, undefined, cbHasLoadedClaimDocsTrue);
      // Add files to the page state
      const tempFiles = [makeFile(), makeFile()];
      await act(async () => {
        await userEvent.upload(
          screen.getByLabelText(/Choose files/),
          tempFiles
        );
      });
      userEvent.click(
        screen.getByRole("button", { name: "Save and continue" })
      );
      await waitFor(() => {
        expect(goToNextPage).toHaveBeenCalledWith(
          { claim },
          {
            claim_id: claim.application_id,
            uploadedAbsenceId: claim.fineos_absence_id,
          }
        );
      });
    });

    describe("when there are previously loaded documents", () => {
      it("renders the FileCardList with documents", () => {
        const cb = (appLogic: AppLogic) => {
          appLogic.documents.hasLoadedClaimDocuments = jest
            .fn()
            .mockReturnValue(true);
          appLogic.documents.documents =
            new ApiResourceCollection<BenefitsApplicationDocument>(
              "fineos_document_id",
              [
                createMockBenefitsApplicationDocument({
                  document_type: DocumentType.identityVerification,
                }),
              ]
            );
        };
        setup(undefined, undefined, cb);
        expect(screen.getByText(/File/)).toBeInTheDocument();
      });

      it("makes API request when there are new files", async () => {
        const claim = new MockBenefitsApplicationBuilder().create();
        const cb = (appLogic: AppLogic) => {
          appLogic.documents.hasLoadedClaimDocuments = jest
            .fn()
            .mockReturnValue(true);
          appLogic.documents.documents =
            new ApiResourceCollection<BenefitsApplicationDocument>(
              "fineos_document_id",
              [
                createMockBenefitsApplicationDocument({
                  document_type: DocumentType.identityVerification,
                }),
              ]
            );
        };
        setup(claim, undefined, cb);
        const tempFiles = [
          makeFile({ name: "file1" }),
          makeFile({ name: "file2" }),
        ];

        await act(async () => {
          await userEvent.upload(
            screen.getByLabelText(/Choose files/),
            tempFiles
          );
        });
        userEvent.click(
          screen.getByRole("button", { name: "Save and continue" })
        );

        await waitFor(() => {
          expect(attach).toHaveBeenCalledWith(
            claim.application_id,
            expect.arrayContaining(expectedTempFiles),
            expect.any(String),
            false
          );
        });
      });

      it("skips an API request if there are no new files added", async () => {
        const cb = (appLogic: AppLogic) => {
          appLogic.documents.hasLoadedClaimDocuments = jest
            .fn()
            .mockReturnValue(true);
          appLogic.documents.documents =
            new ApiResourceCollection<BenefitsApplicationDocument>(
              "fineos_document_id",
              [
                createMockBenefitsApplicationDocument({
                  document_type: DocumentType.identityVerification,
                }),
              ]
            );
        };

        setup(undefined, undefined, cb);

        userEvent.click(
          screen.getByRole("button", { name: "Save and continue" })
        );
        await waitFor(() => {
          expect(attach).not.toHaveBeenCalled();
          expect(goToNextPage).toHaveBeenCalledTimes(1);
        });
      });
    });

    it("renders alert when there is an error loading documents", () => {
      const cb = (appLogic: AppLogic) => {
        appLogic.appErrors = new AppErrorInfoCollection([
          new AppErrorInfo({
            meta: { application_id: "mock_application_id" },
            name: "DocumentsLoadError",
          }),
        ]);
      };
      setup(undefined, undefined, cb);
      expect(
        screen.getByText(
          /An error was encountered while checking your application for documents./
        )
      ).toBeInTheDocument();
    });

    it("When uploading additional docs", async () => {
      const claim = new MockBenefitsApplicationBuilder()
        .medicalLeaveReason()
        .create();
      attach = jest.fn();

      setup(undefined, {
        query: { claim_id: claim.application_id, additionalDoc: "true" },
      });

      userEvent.click(
        screen.getByRole("button", { name: "Save and continue" })
      );

      await waitFor(() => {
        expect(attach).toHaveBeenCalledWith(
          claim.application_id,
          [],
          DocumentType.identityVerification,
          true
        );
      });
    });
  });
});
