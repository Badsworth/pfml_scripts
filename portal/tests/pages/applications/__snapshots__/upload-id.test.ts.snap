// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadId after documents have been loaded when the user doesn't have a Mass ID, renders the page with 'Other' ID content 1`] = `
<div>
  <form
    class="usa-form"
    method="post"
  >
    <h1
      class="js-title margin-top-0 margin-bottom-2 font-heading-sm line-height-sans-3"
      tabindex="-1"
    >
      Your identification
    </h1>
    <div
      class="measure-6"
    >
      <h2
        class="font-heading-lg text-bold"
      >
        Upload an identification document issued by state or federal government
      </h2>
      <div
        class="usa-summary-box margin-bottom-2"
        role="complementary"
      >
        <div
          class="usa-summary-box__body"
        >
          <h3
            class="usa-summary-box__heading font-heading-xs text-bold"
          >
            Document Requirements:
          </h3>
          <div
            class="usa-summary-box__text"
          >
            <ul
              class="usa-list"
            >
              <li>
                The image of the card must be in full color
              </li>
              <li>
                The text must be clear and readable
              </li>
              <li>
                The front and back can be in one file, or in two separate files
              </li>
              <li>
                Each file must be smaller than 4.5 MB
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="border-bottom border-base-light margin-bottom-4 padding-bottom-4"
      >
        <p>
          <strong>
            You can upload a copy of one of the following documents:
          </strong>
        </p>
        <ul
          class="usa-list"
        >
          <li>
            U.S. State or Territory Real ID, both front and back
          </li>
          <li>
            U.S. passport or passport card
          </li>
          <li>
            Permanent Resident Card issued by DHS or INS
          </li>
          <li>
            Employment Authorization Document (EAD) issued by DHS
          </li>
          <li>
            Foreign passport 
            <strong>
              and
            </strong>
             a 
            <a
              href="https://travel.state.gov/content/travel/en/us-visas/employment/temporary-worker-visas.html"
              rel="noopener noreferrer"
              target="_blank"
            >
              work visa
            </a>
          </li>
        </ul>
        <div
          aria-multiselectable="true"
          class="usa-accordion usa-accordion--bordered"
        >
          <h2
            class="usa-accordion__heading"
          >
            <button
              aria-controls="accordion3"
              aria-expanded="false"
              class="usa-accordion__button"
              type="button"
            >
              If you don’t have any of those documents:
            </button>
          </h2>
          <div
            class="usa-accordion__content usa-prose"
            data-testid="accordion3"
            hidden=""
            id="accordion3"
          >
            <p>
              <strong>
                If you don’t have any of those, you can provide one of the following plus proof of your Social Security Number or Individual Tax Identification Number:
              </strong>
            </p>
            <ul
              class="usa-list"
            >
              <li>
                Valid, unexpired U.S. State or Territory License or ID, both front and back
              </li>
              <li>
                Certified copy of a birth certificate filed with a State Office of Vital Statistics or equivalent agency in the individual’s state of birth. (You can only use a 
                <a
                  href="https://prfaa.pr.gov/faqs"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Puerto Rican birth certificate
                </a>
                 if it was issued on or after July 1, 2010.)
              </li>
              <li>
                Certificate of Citizenship (Form N‑560, or Form N‑561)
              </li>
              <li>
                Certificate of Naturalization (Form N‑550 or N‑570)
              </li>
            </ul>
            <p>
              <strong>
                You can provide proof of your Social Security Number using one of the following documents displaying your complete Social Security Number:
              </strong>
            </p>
            <ul
              class="usa-list"
            >
              <li>
                Social Security card
              </li>
              <li>
                W‑2 Form
              </li>
              <li>
                SSA‑1099 Form
              </li>
              <li>
                Non‑SSA‑1099 Form
              </li>
              <li>
                Pay stub with your name on it
              </li>
            </ul>
            <p>
              <a
                href="https://mass.gov/pfmlsubmitinfo"
                rel="noopener noreferrer"
                target="_blank"
              >
                Learn more about verifying your identity with different documents.
              </a>
            </p>
          </div>
        </div>
      </div>
      <details
        class="c-details"
      >
        <summary
          class="margin-bottom-2 text-primary text-underline font-ui-xs"
        >
          Tips for uploading images or PDFs
        </summary>
        <div
          class="margin-bottom-2"
        >
          <h2
            class="font-heading-xs text-bold"
          >
            This website only accepts:
          </h2>
          <ul
            class="usa-list"
          >
            <li>
              PDF documents
            </li>
            <li>
              Images (.jpg, .jpeg, .png)
            </li>
          </ul>
          <h2
            class="font-heading-xs text-bold"
          >
            If your file is not in one of these formats, try one of the following:
          </h2>
          <ul
            class="usa-list"
          >
            <li>
              Take a picture of the document
            </li>
            <li>
              Take a screenshot of the document
            </li>
            <li>
              Save the document as a PDF or image, and try uploading again
            </li>
          </ul>
          <h2
            class="font-heading-xs text-bold"
          >
            If your document is attached to an email:
          </h2>
          <ul
            class="usa-list"
          >
            <li>
              Open the file on your computer or phone
            </li>
            <li>
              Save it as a PDF or image, and try uploading again
            </li>
          </ul>
        </div>
      </details>
      <div
        class="margin-bottom-4"
      >
        <ul
          class="usa-list usa-list--unstyled measure-5"
        />
        <ul
          class="usa-list usa-list--unstyled measure-5"
        />
        <label
          class="margin-top-2 usa-button usa-button--outline"
        >
          Choose files
          <input
            accept="image/*,.pdf"
            class="c-file-card-list__input"
            multiple=""
            type="file"
          />
        </label>
      </div>
    </div>
    <button
      class="usa-button position-relative margin-top-4"
      type="submit"
    >
      Save and continue
    </button>
  </form>
</div>
`;
